File: 010.数据流的中位数/main.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

class MedianFinder {
private:
    priority_queue<int, vector<int>, less<>> queMin;
    priority_queue<int, vector<int>, greater<>> queMax;


public:
    MedianFinder()
    = default;

    void addNum(int num) {
        if (queMin.empty() || num < queMin.top()) {
            queMin.push(num);
            if (queMax.size() + 1 < queMin.size()) {
                queMax.push(queMin.top());
                queMin.pop();
            }
        } else {
            queMax.push(num);
            if (queMax.size() > queMin.size()) {
                queMin.push(queMax.top());
                queMax.pop();
            }
        }
    }

    double findMedian() {
        if (queMin.size() > queMax.size()) {
            return queMin.top();
        }
        return (queMin.top() + queMax.top()) / 2.0;
    }
};

int main(int argc, char const *argv[]) {
    MedianFinder medianFinder{};
    medianFinder.addNum(1);                    // arr = [1]
    medianFinder.addNum(2);                    // arr = [1, 2]
    cout << medianFinder.findMedian() << endl; // 返回 1.5 ((1 + 2) / 2)
    medianFinder.addNum(3);                    // arr[1, 2, 3]
    cout << medianFinder.findMedian() << endl; // return 2.0
    return 0;
}

File: 010.数据流的中位数/main.js
var MedianFinder = function () {
  this.arr = [];
};

/**
 * @param {number} num
 * @return {void}
 */
MedianFinder.prototype.addNum = function (num) {
  this.arr.push(num);
  this.arr = this.arr.sort();
};

/**
 * @return {number}
 */
MedianFinder.prototype.findMedian = function () {
  const size = this.arr.length;
  if (size % 2 != 0) {
    // 返回中间的数字
    const middleIndex = size / 2;
    return this.arr[middleIndex];
  } else {
    // 返回中间两个数字的平均值。
    const afterIndex = size / 2;
    const beforeIndex = afterIndex - 1;
    return this.arr[afterIndex] + this.arr[beforeIndex];
  }
};
